File: python/soft_clustering.py
# soft clustering
import math
import random

# K-means clustering (soft) via Expectation-Maximization (EM) algorithm
# soft assignment based on what each data point’s responsibility for each cluster
# given n data points (Data1 . . . Datan) and k centers (x1 . . . xk), we need to construct a k × n 

# centers to clusters
def expect(HM, centers, points, stiffness, m):
    totals = [0 for _ in range(len(points))]
    # expectation
    # calculate force for each point and center
    for i in range(len(centers)):
        for j in range(len(points)):
            distance =  math.sqrt(sum((centers[i][k] - points[j][k])**2 for k in range(m)))
            force = math.pow(math.e, -stiffness*distance)
            HM[i][j] = force
    # get sum force for each point
    for i in range(len(centers)):
        for j in range(len(points)):
            totals[j] += HM[i][j]
    # divide HM values by sum
    for i in range(len(centers)):
        for j in range(len(points)):
            HM[i][j] /= totals[j]

# clusters to centers
def maximize(k, points, HM, m):
    # x i,j = Hidden Matrix i dot Data j / Hidden Matrix i dot 1 (aka sum of row i)
    # x i,j where i is the center and j is the datapoint, and HM[i] is the row vector
    # pythonic dot product: sum(i*j for (i, j) in zip(list1, list2)) or a = A @ B yay !
    new_centers = [[0 for _ in range(m)] for _ in range(k)]
    for i in range(k):
        for j in range(m):
            new_centers[i][j] = round(sum(a*b for (a, b) in zip(HM[i], [point[j] for point in points])) / sum(HM[i]),3)
    return new_centers

def soft_clustering(k, m , stiffness, points):
    # expect
    HM = [[1/k for _ in range(len(points))] for _ in range(k)] # H[i][j] = responsibiliy of center i on point j
                                                            # HM = Forcei,j / Sum of all center's forces
                                                             # Forcei,j = e^(-stiffness*distance(dataj, centeri))
    # centers = points[:k]
    centers = random.sample(points, k)
    i = 0
    while i <= 20:
        expect(HM, centers, points, stiffness, m)
        new_centers = maximize(k, points, HM, m)
        if new_centers == centers:
            break
        centers = new_centers
        i += 1
    # for row in HM:
    #     print(row)
    return centers

# centers = soft_clustering(3, 2, 0.5,[[1,2], [2,3], [4,5], [3,2], [9,0], [0,0], [0,7.64], [10, 294.4], [9.0, 299.4]])
# for center in centers:
#     print(center)

File: soft_clustering.js
function expect(HM, centers, points, stiffness){
    let totals = Array.from({length: points.length}, ()=>0)
    for(let i=0; i<centers.length; i++){
        for (j=0; j<points.length; j++){
            distance =  fn_distance(points[j], centers[i])
            force = Math.pow(Math.E, -stiffness*distance)
            HM[i][j] = force
        }
    }
    for(let i=0; i<centers.length; i++){
        for(let j=0; j<points.length; j++){
            totals[j] += HM[i][j]
        }
    }

    for(i=0; i<centers.length; i++){
        for(j=0; j<points.length; j++){
            HM[i][j] /= totals[j]
        }
    }
    return HM
}

function maximize(k, points, HM){
    let newCenters = []
    for(let i=0; i<k; i++){
        newCenters.push([0,0])
    }
    for (let i=0; i<k; i++) {
        row = HM[i];      
        rowSum = row.reduce((a, b) => a + b)

        for (let j=0; j<2; j++) {
            let weighted = 0
            for (let p=0; p <points.length; p++) {
                weighted += row[p] * points[p][j]
            }
            newCenters[i][j] = (weighted / rowSum)
        }
    }
    return newCenters;
}

function soft_clustering(k, stiffness){
    let HM = Array.from({ length: k }, () =>
        Array(points.length).fill(1 / k)
    )
    let centers = initializePoints(points, k)
    let i = 0
    while(i <= 20){
        HM = expect(HM, centers, points, stiffness)
        let newCenters = maximize(k, points, HM)
        if(newCenters === centers){
            break
        }
        centers = newCenters
        i += 1
    }
    // console.log(HM)
    softAssignColors(HM, k)
    return centers
}

function softAssignColors(HM, k){
    points.forEach((point, p) => {
        circle = svg.getElementById(`point-${point}`)
        color = [0, 0, 0, 1]
        for(let i=0; i<k; i++){
            t = HM[i][p]
            color = interpolateColors(color, colors[i], t)
        }
        circle.setAttributeNS(null, 'fill', `rgb(${color[0]},${color[1]},${color[2]},${color[3]})`)
    })
}

// interpolate rgb colors
function interpolateColors(color, newColor, t){
    recolor = [0,0,0,1]
    for(let i=0; i<3; i++){
        recolor[i] = (color[i] * (1-t)) + (newColor[i] * t)
    }
    return recolor
}

// console.log(soft_clustering(3, 0.5, [[1,2], [2,3], [4,5], [3,2], [9,0], [0,0], [0,7.64], [10, 294.4], [9.0, 299.4]]))